typing/typecore.ml

```diff
@@ -5109,6 +5168,7 @@ and type_function ?in_function
       Final_arg { partial_mode = Alloc_mode.join [arg_mode; alloc_mode] }
     end
   in
+  let cases_expected_mode = expect_mode_cross env ty_res cases_expected_mode in 
```

Comment:
This can actually only have an effect when `uncurried_function` is `false`. We can save a little bit of work by moving this into the `else` branch above.